### PR TITLE
fix: next round of restructuring.

### PR DESCRIPTION
**Pull request by andrewferlitsch**
**REQUIRED:** Add a summary of your PR here, typically including why the change is needed and what was changed. Include any design alternatives for discussion purposes.

<br>
Moved helper functions into abstract class.
Changed TW rule parsing functions into NotebookRule classes.
Additional changes for TW needs

Potential future code refinements:
1. handling of the two global variables: last_tag and num_errors
2. consolidating error codes and msgs into a dictionary style structure.
3.  undecided on how to or change generating the index -- still a standalone Python function.
<br><br><br>

**REQUIRED:** Fill out the below checklists or remove if irrelevant
1. If you are opening a PR for `Official Notebooks` under the [notebooks/official](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/notebooks/official) folder, follow this mandatory checklist:
- [ ] Use the [notebook template](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/notebook_template.ipynb) as a starting point.
- [ ] Follow the style and grammar rules outlined in the above notebook template.
- [ ] Verify the notebook runs successfully in Colab since the automated tests cannot guarantee this even when it passes.
- [ ] Passes all the required automated checks. You can locally test for formatting and linting with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).
- [ ] You have consulted with a tech writer to see if tech writer review is necessary. If so, the notebook has been reviewed by a tech writer, and they have approved it.
- [ ] This notebook has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/CODEOWNERS) file under the `Official Notebooks` section, pointing to the author or the author's team.
- [ ] The Jupyter notebook cleans up any artifacts it has created (datasets, ML models, endpoints, etc) so as not to eat up unnecessary resources.

<br>

2. If you are opening a PR for `Community Notebooks` under the [notebooks/community](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/notebooks/community) folder:
- [ ] This notebook has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/community/CODEOWNERS) file under the `Community Notebooks` section, pointing to the author or the author's team.
- [ ] Passes all the required formatting and linting checks. You can locally test with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).

<br>

3. If you are opening a PR for `Community Content` under the [community-content](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/community-content) folder:
- [ ] Make sure your main `Content Directory Name` is descriptive, informative, and includes some of the key products and attributes of your content, so that it is differentiable from other content
- [ ] The main content directory has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/community-content/CODEOWNERS) file under the `Community Content` section, pointing to the author or the author's team.
- [ ] Passes all the required formatting and linting checks. You can locally test with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).

### CODE REVIEW

File: notebooks/notebook_template_review.py [hunk 1]
```diff
@@ -193,71 +194,81 @@ def parse_notebook(path: str) -> None:
         
         cells = content['cells']
         
-        NotebookRule.init()
+        NotebookRule.init(path, content['cells'])
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:197]: You can omit `init`, just do `NotebookRule(...)`

File: notebooks/notebook_template_review.py [hunk 2]
```diff
@@ -193,71 +194,81 @@ def parse_notebook(path: str) -> None:
...
         
-        CopyrightRule().validate(path, cells)
+        CopyrightRule().validate()
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:199]: You should instantiate outside this method and just call them here.

File: notebooks/notebook_template_review.py [hunk 3]
```diff
@@ -268,73 +279,112 @@ def init():
...
+    def get_cell(self) -> list:
+        """
+            Get the next notebook cell.
+
+            Returns:
+
+            cell: content of the next cell
+        """
+        while self.empty_cell():
+            NotebookRule.cell_index += 1
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:294]: You are modifying the class definition, instead of the actual instance of the class. 

This needs to be `self.cell_index`.

Better yet, remove `cell_index` altogether.

File: notebooks/notebook_template_review.py [hunk 4]
```diff
@@ -268,73 +279,112 @@ def init():
         NotebookRule.costs = ''
     
     @abstractmethod
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, cells: list) -> None:
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:282]: The definition has a `cells` parameter but nothing is passed in when it's called.

Comment (ivanmkc) [reply notebooks/notebook_template_review.py:282]: The subclasses don't conform properly to this abstractmethod.

File: notebooks/notebook_template_review.py (hunk 1 above)
Comment (ivanmkc) [reply notebooks/notebook_template_review.py:197]: 1. NotebookRule is an `abstractclass` so should not be instantiated directly (instantiate subclasses of it).
2. Please refactor to pass in cells at the `validate` method. `init` method should have parameters related to the rule definition itself.
3. Pass in cells to the `validate` method.

File: notebooks/notebook_template_review.py (hunk 3 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:294]: I use the static version of 'cell_index' instead of the instance version to maintain tracking where I am in the notebook. If I used the instance version, I believe the location would be lost when the next Rule is executed.

File: notebooks/notebook_template_review.py (hunk 4 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:282]: yup, missed that

File: notebooks/notebook_template_review.py (hunk 2 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:199]: not sure what the comment means. Can you give an example?

File: notebooks/notebook_template_review.py (hunk 3 above)
Comment (ivanmkc) [reply notebooks/notebook_template_review.py:294]: I don't think you need `get_cell`, just call pass in cells to `validate` and loop over them.

File: notebooks/notebook_template_review.py (hunk 1 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:197]: I tried that. Got an error about initializing (__init__) an abstract class. Hence why I went the static method route.

File: notebooks/notebook_template_review.py [hunk 5]
```diff
@@ -193,71 +194,81 @@ def parse_notebook(path: str) -> None:
...
+    colab_link = None  # The Colab link
+    workbench_link = None  # The Workbench link
+    desc = ''  # The description in the objective section
+    uses = ''  # The resources used in the objective section
+    steps = ''  # The steps in the objective section
+    costs = []  # The paid services used as derived from the resources used subsection
     
     @staticmethod
-    def init():
+    def init(path: str,
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:261]: You shouldn't need to define an `init` method in your abstractclass. The subclasses will define their own.

Just describe the methods in your ABC.

**Review by ivanmkc — CHANGES_REQUESTED** [review]

**Review by gericdong — APPROVED** [review]
LGTM

File: notebooks/notebook_template_review.py (hunk 3 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:294]: I see two approaches. One we can retain location knowledge in the abstract base class (current). The other is to retain location knowledge outside of the class. This would require additional tracking code between the rules. While I can see the benefit in that, your example of using a generator (prior PR) would be very messy.

The third choice would make finding the cells location independent. Then the derived case could ask the abstract class to go find the cell. BUT, you would loose cell order dependence. Perhaps one could make a dictionary of call order after the rules have been applied, and then do a pass over the dictionary checking order.

The third approach (I used to be a compiler writer early in my career) would be my preference. But, that's a lot of engineering vs the benefit of coding the most elegant solution.

File: notebooks/notebook_template_review.py (hunk 1 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py]: done

File: notebooks/notebook_template_review.py (hunk 3 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py]: done

File: notebooks/notebook_template_review.py (hunk 4 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py]: done

File: notebooks/notebook_template_review.py (hunk 5 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py]: done

File: notebooks/notebook_template_review.py [hunk 6]
```diff
@@ -193,148 +198,141 @@ def parse_notebook(path: str) -> None:
...
-        NotebookRule.desc = ''
-        NotebookRule.uses = ''
-        NotebookRule.steps = ''
-        NotebookRule.costs = ''
-    
+    path = None     # The relative path to the notebook
+
     @abstractmethod
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, cells: list) -> list:
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:258]: Can you add a comment on what this does?

Should be:

```
def validate(self, cells: list) -> List[SomeTypeHere]:
```

File: notebooks/notebook_template_review.py [hunk 7]
```diff
@@ -193,148 +198,141 @@ def parse_notebook(path: str) -> None:
...
-        else:
-            NotebookRule.cell_index = cell_index - 1
+            cells.pop(0)
+        return cells
 
 
-class TitleRule(NotebookRule):
-    def validate(self, path: str, cells: list) -> None: 
+class TitleRule(NotebookRule): 
+    def validate(self, cells: list) -> (list, str):
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:311]: Method signature has to match the abstractclassethod. Considering using mypy to help you find these issues.

File: notebooks/notebook_template_review.py [hunk 8]
```diff
@@ -175,14 +176,18 @@ def parse_dir(directory: str) -> None:
...
-            parse_notebook(entry.path)
+            parse_notebook(entry.path, tag=directory.split('/')[-1], linkback=None)
 
 
-def parse_notebook(path: str) -> None:
+def parse_notebook(path: str,
+                   tag: str,
+                   linkback: str) -> None:
     """
         Review the specified notebook.
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:186]: Can you elaborate: what does "Review" do?

File: notebooks/notebook_template_review.py [hunk 9]
```diff
@@ -175,14 +176,18 @@ def parse_dir(directory: str) -> None:
...
             parse_dir(entry.path)
         elif entry.name.endswith('.ipynb'):
-            parse_notebook(entry.path)
+            parse_notebook(entry.path, tag=directory.split('/')[-1], linkback=None)
 
 
-def parse_notebook(path: str) -> None:
+def parse_notebook(path: str,
+                   tag: str,
+                   linkback: str) -> None:
```
[the remark this answers is not in the document]
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:184]: done

File: notebooks/notebook_template_review.py (hunk 6 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py]: added

File: notebooks/notebook_template_review.py [hunk 10]
```diff
@@ -193,148 +198,141 @@ def parse_notebook(path: str) -> None:
...
+                     errmsg: str) -> None:
+        """
+        Report an error.
+            If args.errors_codes set, then only report these errors. Otherwise, all errors.
+
+        code: The error code number.
+        errmsg: The error message
+        """
+
+        global num_errors
```
[the remark this answers is not in the document]
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py]: num_errors not global anymore

File: notebooks/notebook_template_review.py (hunk 7 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py]: was updated

File: notebooks/notebook_template_review.py (hunk 8 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:190]: updated

File: notebooks/notebook_template_review.py [hunk 11]
```diff
@@ -154,268 +154,324 @@ class ErrorCode(Enum):
...
     """
-    with open(path, 'r') as f:
-        try:
-            content = json.load(f)
-        except:
-            print("Corrupted notebook:", path)
-            return
-        
-        cells = content['cells']
+    notebook = Notebook(path)
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:199]: The rules don't need to be instantiated for every notebook. Just create them once in the global scope.

File: notebooks/notebook_template_review.py [hunk 12]
```diff
@@ -154,268 +154,324 @@ class ErrorCode(Enum):
...
+    installation = InstallationRule()
+    restart = RestartRule()
+    versions = VersionsRule()
+    beforebegin = BeforeBeginRule()
+    enableapis = EnableAPIsRule()
+    setupproject = SetupProjectRule()
+
         
-        NotebookRule.init()
+    copyright.validate(notebook)
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:220]: The point of abstracting to Rule's is that you don't have to call each instance separately.

Just do:

for rule in rules:
   rule.validate(notebook)
   
You probably want to return something from the validate method and do something with it.

File: notebooks/notebook_template_review.py [hunk 13]
```diff
@@ -154,268 +154,324 @@ class ErrorCode(Enum):
...
+    helpers.validate(notebook)
+    installation.validate(notebook)
+    restart.validate(notebook)
+    versions.validate(notebook)
+    beforebegin.validate(notebook)
+    enableapis.validate(notebook)
+    setupproject.validate(notebook)
+
+    
+    if objective.desc != '':
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:239]: This seems unrelated to validation and seems to be a side-effect for index creation.

File: notebooks/notebook_template_review.py [hunk 14]
```diff
@@ -154,268 +154,324 @@ class ErrorCode(Enum):
...
+        )
         
-        CopyrightRule().validate(path, cells)
-            
-        NoticesRule().validate(path, cells)
+    return notebook.num_errors
+
+class Notebook(object):
+    '''
+    Class for navigating thru a notebook
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:256]: through?

File: notebooks/notebook_template_review.py [hunk 15]
```diff
@@ -154,268 +154,324 @@ class ErrorCode(Enum):
...
-        NotebookRule.colab_link = None
-        NotebookRule.workbench_link = None
-        NotebookRule.desc = ''
-        NotebookRule.uses = ''
-        NotebookRule.steps = ''
-        NotebookRule.costs = ''
-    
     @abstractmethod
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, notebook: Notebook):
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:352]: Should return bool.

File: notebooks/notebook_template_review.py (hunk 13 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:239]: Yes, it is for indexing

File: notebooks/notebook_template_review.py [hunk 16]
```diff
@@ -154,268 +154,324 @@ class ErrorCode(Enum):
...
-
-        path: used only for reporting an error
-        cells: The content cells (JSON) for the notebook
         """
-        cell, cell_index = get_cell(path, cells, NotebookRule.cell_index)
+        cell = notebook.peek()
         if not cell['source'][0].startswith('# '):
-            report_error(path, ErrorCode.ERROR_TITLE_HEADING, "title cell must start with H1 heading")
-            NotebookRule.title = ''
+            notebook.report_error(ErrorCode.ERROR_TITLE_HEADING, "title cell must start with H1 heading")
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:388]: Instead of storing state in a notebook object, it's much simpler to just loop over notebook cells and return results.
The less places you store state (especially, mutable state like this), the better.

File: notebooks/notebook_template_review.py [hunk 17]
```diff
@@ -478,443 +534,339 @@ def validate(self, path: str, cells: list) -> None:
...
-    cell = cells[cell_index]
-    if cell['cell_type'] == 'markdown':
-        check_text_cell(path, cell)
-    return cell, cell_index + 1
+            for mistake, brand in branding.items():
+                if mistake in line:
+                    notebook.report_error(ErrorCode.ERROR_TWRULE_BRANDING, f"Branding {mistake} -> {brand}: {line}")
 
+class SentenceCaseTWRule(NotebookRule):
+    def validate(self,
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:819]: The `validate` method must remain the same signature as defined in `NotebookRule` or else it's meaningless.

File: notebooks/notebook_template_review.py (hunk 12 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:220]: done

File: notebooks/notebook_template_review.py (hunk 14 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:256]: done

File: notebooks/notebook_template_review.py (hunk 15 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:352]: done

File: notebooks/notebook_template_review.py (hunk 11 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:199]: done

File: notebooks/notebook_template_review.py (hunk 13 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:239]: added comments to code

File: notebooks/notebook_template_review.py (hunk 17 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:872]: done

File: notebooks/notebook_template_review.py [hunk 18]
```diff
@@ -154,268 +157,354 @@ class ErrorCode(Enum):
...
+    notebook = Notebook(path)
+    
+    # Cell Validation
+    rules = [ copyright, notices, title, links, overview, objective,
+              recommendations, dataset, costs, setuplocal, helpers,
+              installation, restart, versions, beforebegin, enableapis,
+              setupproject
+            ]
+    
+    for rule in rules:
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:211]: This looks better.

File: notebooks/notebook_template_review.py [hunk 19]
```diff
@@ -154,268 +157,354 @@ class ErrorCode(Enum):
...
-            content = json.load(f)
-        except:
-            print("Corrupted notebook:", path)
-            return
-        
-        cells = content['cells']
+    notebook = Notebook(path)
+    
+    # Cell Validation
+    rules = [ copyright, notices, title, links, overview, objective,
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:205]: Ideally create these outside the method and pass them in as a parameter so your method is reusable for different rules.

File: notebooks/notebook_template_review.py [hunk 20]
```diff
@@ -154,268 +157,354 @@ class ErrorCode(Enum):
...
-    def validate(self, path: str, cells: list) -> None:
-        pass
+    def validate(self, notebook: Notebook) -> bool:
+        '''
+        Applies cell specific rules to validate whether the cell 
+        does or does not conform to the rules.
+        
+        Returns whether the cell passed the validation rules
+        '''
+        return False
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:366]: This is an abstract method, so should be `pass`

File: notebooks/notebook_template_review.py [hunk 21]
```diff
@@ -426,495 +515,446 @@ def validate(self, path: str, cells: list) -> None:
...
-    notebook: The notebook path.
-    code: The error code number.
-    errmsg: The error message
-    """
-    
-    global num_errors
+
+
+class TextTWRule(TextRule):
+    def validate(self, notebook: Notebook, text: List[str]) -> bool:
```
Comment (ivanmkc) [inline notebooks/notebook_template_review.py:909]: Signature not conforming to abstractclass.

Comment (ivanmkc) [reply notebooks/notebook_template_review.py:909]: That means that this method is never called and your abstract classes default method is called instead.

**Review by ivanmkc — COMMENTED** [review]
It looks good except the `Notebook.get`, `Notebook.pop`, `Notebook.peek` logic.

Each rule is independent, but the way the code is written now, there is shared mutable state stored in the Notebook._cell_index parameter. We can easily remove this parameter and have each rule check the cell that they need.

If a rule depends on the first cell:

```
Just validate on `notebooks.cells[0]["source"]`
```

File: notebooks/notebook_template_review.py (hunk 18 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:211]: thanks

File: notebooks/notebook_template_review.py (hunk 20 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:366]: done

File: notebooks/notebook_template_review.py (hunk 21 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:909]: This is derived from the TextRule abstract class, not the general NotebookRule class. It assumes that instead of passing a cell, it is passed some passage (list) of text strings.

File: notebooks/notebook_template_review.py (hunk 19 above)
Comment (andrewferlitsch) [reply notebooks/notebook_template_review.py:205]: done
